### ISSUE TITLE
Question about Souls series item injection vulnerabilities

### ISSUE THREAD

**Issue by admiralzeech** [Jellybaby34/Dark-Souls-3-PvP-Watchdog#75]
Apologies if this is not quite the right place to ask this, but I wanted to get some insight into what vulnerabilities allow the item injection (and applying curse status etc) hacks to occur in the Souls series.

I'm a rusty former games programmer (although mostly gameplay logic and rarely anything to do with networking), so I have a basic understanding.  "Back in my day" when you sent a network message, the other side would only process valid messages of which there were only a limited set of possibilities.

At first glance, it seems the item injection / status injection hacks in Souls games use available functionality rather hacky stuff like buffer overruns and such?

So how is it possible to inject items into another client's inventory, teleport them, or apply invalid statuses to them?  Is it because Souls games allow other clients to perform Remote Procedure Calls on your client without any restrictions or sanity checks?

**Comment by LukeYui** [comment]
It's a lot more simple than that.

The souls games are coded in such a way that packet source verification checks are often incredibly straightforward.

An example of this is applying effects to others. All that needs to be done here is to send a normal networking packet that would say "I am player ID 2 applying effect 100" and switch it to "I am player ID 1 applying effect 100". Even if the local player who receives this packet has ID 1, when the game sees that packet it will apply it to them instead of me.

This is the same for every single crash and exploit in the souls games:
- Packets that have been packed can be exploited to overrun their buffer.
- Packets that access arrays can be given arguments that cause out-of-bounds crashes.
- Packets that allocate memory can be given arguments to corrupt the heap.
- The list goes on

Item injection specifically is, as you say, a normal packet. This is how item packets work:

- Remote player 2 drops an item, this sends an "Item Drop" case item packet to the host.
- The host receives that packet, and creates the item instance in the world.
- Remote player 3 picks up the item, this sends an "Item pickup request" case packet to the host.
- The host sends back an "Item delete" case item packet to the session which deletes the item instance.
- The host also sends an "Item give" case item packet which gives the item to the remote player 3.

I think you see where this is going. It's possible to just simply send the "Item give" case item packet directly, there are no checks in the base game to prevent this. I hope this clears things up.

tl;dr - Every exploitable packet uses normal game functions, but the packet data is tweaked in such a way that it results in strange or undefined behaviour.

**Comment by admiralzeech** [comment]
Thanks for the reply!

It certainly is depressing how they can go for 4 games without applying even the most basic of sanity checks in their network security...
